### PR TITLE
Fix OpenAI tool result ids

### LINKED ISSUE #2019: OpenAI converter can emit empty tool_call_id for missing functionResponse.id
## Summary

Laminar's OpenAI Chat Completions converter can emit an invalid tool-result message with an empty `tool_call_id` when an internal `functionResponse.id` is missing.

This breaks OpenAI-compatible endpoints because `role: "tool"` messages require a non-empty `tool_call_id` that matches a previous assistant `tool_calls[].id`.

## What we observed

Signals requests using the OpenAI-compatible path can fail when the converted request contains a tool result like:

```json
{
  "role": "tool",
  "tool_call_id": "",
  "content": "..."
}
```

The recent LAM-1771 fix correctly avoids forwarding `reasoning_effort` when tools are present, and the converter also handles missing assistant tool-call ids by generating a UUID. But the matching tool-result path still falls back to an empty string.

## Source-level cause

In `app-server/src/llm/openai/conversions.rs`, assistant tool calls are repaired when `functionCall.id` is missing:

```rust
let id = fc.id.unwrap_or_else(|| uuid::Uuid::new_v4().to_string());
```

But tool results use an empty fallback when `functionResponse.id` is missing:

```rust
let tool_call_id = fr.id.unwrap_or_default();
```

That creates invalid OpenAI messages if the agent/tool execution layer emits a Gemini-shaped `functionResponse` without carrying the original call id.

## Why this can happen

Laminar's internal provider schema models both ids as optional:

```rust
pub struct ProviderFunctionCall {
    pub id: Option<String>,
    pub name: String,
    pub args: Option<serde_json::Value>,
}

pub struct ProviderFunctionResponse {
    pub id: Option<String>,
    pub name: String,
    pub response: serde_json::Value,
}
```

That is compatible with Gemini-style content, but OpenAI Chat Completions requires the tool-result id to be non-empty and tied to the previous assistant tool call.

## Expected behavior

The OpenAI conversion path should never emit `tool_call_id: ""`.

Possible fixes:

1. Preserve generated fallback ids across the assistant `function_call` and the matching later `function_response`.
2. If preservation is not possible at conversion time, reject the malformed internal request before sending it to OpenAI-compatible endpoints with a clear error.
3. If there is exactly one preceding unresolved tool call, infer that id for a missing `functionResponse.id`; otherwise fail fast to avoid mis-associating tool results.

## Impact

This can keep Signals / agent investigations from running against OpenAI-compatible gateways even after the `reasoning_effort + tools` fix is present.

### PR DESCRIPTION
**Pull request by herculesggimenes**
## Summary
- track emitted OpenAI tool call ids while converting Laminar provider messages
- reuse the pending tool call id when a following `functionResponse.id` is missing
- cover explicit and generated tool call ids so the converter does not emit `tool_call_id: ""`

Fixes #2019.

## Test
```bash
cargo test --manifest-path app-server/Cargo.toml llm::openai::conversions::tests
```

---

> [!NOTE]
> **Medium Risk**
> Changes LLM request shaping for tool-use flows; wrong pairing could break OpenAI API calls or drop tool results, though scope is limited to the OpenAI converter with broad test coverage.
> 
> **Overview**
> Fixes OpenAI Chat Completions request conversion so **`role: "tool"`** messages no longer get an empty **`tool_call_id`** when internal **`function_response`** parts omit **`id`**.
> 
> The converter now keeps a **FIFO queue** of tool calls already emitted in the same request (id + function name). Each **`function_response`** resolves its **`tool_call_id`** via **`resolve_tool_call_id`**: match an explicit id against pending entries (and remove it), or pair by function name in order. Unmatched responses are **dropped** instead of sent with **`""`** (e.g. response before any call, or name mismatch).
> 
> Generated UUIDs on assistant **`function_call`** parts are queued the same way, so a later response without id still reuses that id. New unit tests cover FIFO pairing for duplicate tool names, explicit-id ordering, and stale-pending edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6507a190ec975e7c00abe59dba0307c66bce86a4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by herculesggimenes** [comment]
@greptileai review the latest changes.

File: app-server/src/llm/openai/conversions.rs [hunk 1]
```diff
@@ -154,7 +160,17 @@ fn append_content_as_messages(content: &ProviderContent, out: &mut Vec<Value>) {
             continue;
         }
         if let Some(fr) = part.function_response {
-            let tool_call_id = fr.id.unwrap_or_default();
+            let tool_call_id = fr.id.unwrap_or_else(|| {
+                pending_tool_call_ids
+                    .pop_front()
+                    .unwrap_or_else(|| uuid::Uuid::new_v4().to_string())
```
[the remark this answers is not in the document]
Comment (herculesggimenes) [reply app-server/src/llm/openai/conversions.rs]: Fixed in 848a263b. Missing-id function responses no longer create a UUID or consume a stale pending id when there is no unambiguous matching call. The converter now skips a missing-id tool response if it appears before any matching function call, so it will not emit an orphan `role: "tool"` message with an invalid `tool_call_id`. Added `missing_tool_response_id_before_tool_call_is_skipped` to cover this case.

File: app-server/src/llm/openai/conversions.rs [hunk 2]
```diff
@@ -204,6 +223,39 @@ fn append_content_as_messages(content: &ProviderContent, out: &mut Vec<Value>) {
...
+        return pending_tool_calls.pop_front().map(|pending| pending.id);
+    }
+
+    let matching_positions: Vec<usize> = pending_tool_calls
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, pending)| (pending.name == response_name).then_some(idx))
+        .collect();
+    if matching_positions.len() == 1 {
+        return pending_tool_calls
```
[the remark this answers is not in the document]
Comment (herculesggimenes) [reply app-server/src/llm/openai/conversions.rs]: Fixed in 0c59bd4b. Same-name missing-id tool results now pair FIFO when the front pending call has the same function name, which covers the normal ordered case of calling the same tool multiple times and receiving id-less results in order. The converter still avoids unsafe pairing when the matching duplicate-name calls are not at the front of the queue; that ambiguous case is skipped rather than assigned to an unrelated older call. Added `same_name_missing_tool_response_ids_are_paired_fifo` and `missing_tool_response_id_for_non_front_duplicate_name_is_skipped`.

**Comment by herculesggimenes** [comment]
@greptileai review the latest changes.

File: app-server/src/llm/openai/conversions.rs [hunk 3]
```diff
@@ -204,6 +223,46 @@ fn append_content_as_messages(content: &ProviderContent, out: &mut Vec<Value>) {
...
+            .remove(matching_positions[0])
+            .map(|pending| pending.id);
+    }
+
+    None
```
[the remark this answers is not in the document]
Comment (herculesggimenes) [reply app-server/src/llm/openai/conversions.rs]: Fixed in 8133b197. Missing-id responses now consume the first pending call with the same function name, so duplicate same-name calls are paired FIFO even when those duplicate calls are behind another pending tool. Added `missing_tool_response_id_uses_first_matching_tool_name` for this case.

**Comment by herculesggimenes** [comment]
@greptileai review the latest changes.

File: app-server/src/llm/openai/conversions.rs [hunk 4]
```diff
@@ -166,6 +181,10 @@ fn append_content_as_messages(content: &ProviderContent, out: &mut Vec<Value>) {
         }
         if let Some(fc) = part.function_call {
             let id = fc.id.unwrap_or_else(|| uuid::Uuid::new_v4().to_string());
+            pending_tool_calls.push_back(PendingToolCall {
+                id: id.clone(),
+                name: fc.name.clone(),
+            });
```
[the remark this answers is not in the document]
Comment (herculesggimenes) [reply app-server/src/llm/openai/conversions.rs:187]: Fixed in faad5901. Explicit-id tool results that arrive before their matching function call are no longer emitted as orphan tool messages or allowed to leave a stale pending call behind. The converter now records early explicit result ids; when the later function call with that id appears, it is not queued for future missing-id responses. Added `early_explicit_tool_response_does_not_leave_stale_pending_call` to cover the stale-pending scenario.

**Comment by herculesggimenes** [comment]
@greptileai review the latest changes.

File: app-server/src/llm/openai/conversions.rs [hunk 5]
```diff
@@ -166,6 +189,12 @@ fn append_content_as_messages(content: &ProviderContent, out: &mut Vec<Value>) {
         }
         if let Some(fc) = part.function_call {
             let id = fc.id.unwrap_or_else(|| uuid::Uuid::new_v4().to_string());
+            if !early_tool_result_ids.remove(&id) {
+                pending_tool_calls.push_back(PendingToolCall {
+                    id: id.clone(),
+                    name: fc.name.clone(),
+                });
```
[the remark this answers is not in the document]
Comment (herculesggimenes) [reply app-server/src/llm/openai/conversions.rs]: Fixed in 6507a190. I removed the early-result-id suppression path. If an explicit result arrives before its matching call, that early orphan result is skipped because no pending call exists yet, but the later function call is still queued normally. A following id-less result with the same function name can then pair to that pending call, so the assistant tool call is not left without a matching `role: "tool"` message. The existing `early_explicit_tool_response_does_not_leave_stale_pending_call` test now asserts that behavior.

**Comment by herculesggimenes** [comment]
@greptileai review the latest changes.

**Comment by Rainhunter13** [comment]
@herculesggimenes 

To confirm my understanding — this happens when using OpenAI compatible gateway with a provider that doesn't return function call ids in the response? Which fails in the next step cause function call/response id is required by the OpenAI compatible gateway.

LGTM in general, but would like to confirm issue scope (what providers/gateways) so I can try to reproduce, thanks

**Comment by herculesggimenes** [comment]
We use AWS Bedrock through our internal OpenAI-compatible gateway. I checked our gateway path, and it preserves IDs when they exist:
- Bedrock/Anthropic `tool_use.id` is returned as OpenAI `tool_calls[].id`
- OpenAI `tool_call_id` is forwarded back as Bedrock/Anthropic `tool_result.tool_use_id`

The issue seems to be in the conversion from Laminar’s internal provider format back into OpenAI Chat Completions messages.

Laminar’s internal format is Gemini-like, where a tool response is represented as a `functionResponse`. In that shape, `functionResponse.id` can be missing.

So the repro is:

1. Build a Laminar internal `ProviderRequest` with a prior `functionCall` that has an id, or one where the OpenAI converter generates an id.
2. Add a following `functionResponse` for the same tool name, but with `id: None`.
3. Convert that request to OpenAI Chat Completions format.

**Comment by Rainhunter13** [comment]
@herculesggimenes, thanks for the details

You mentioned that your gateway returns ids as OpenAI `tool_calls[].id` and reads as `tool_call_id`. I believe this is exactly fields we parse and send when converting from/to OpenAI format, at lines 165 and 268. It also works in my tests with OpenAI API.

Given your gateway uses OpenAI format, I am struggling to understand how empty tool id can appear in the the first place, so we need to handle it with some fifo fallback. We do use Gemini format internally, but conversion between OpenAI and Gemini preserves function ids in both calls and responses.

Is there any chance that your gateway doesn't return tool call ids when requesting tool calls, but still expects them in the next request?

**Comment by Rainhunter13** [comment]
One error I can see for sure is that when there is a single tool call with empty id, we should re-send it as empty instead of generating a new uuid. But that fixes the case only when there's single missing tool call id.

UPD: This is now fixed and released in latest images.

**Comment by herculesggimenes** [comment]
Thank you! I will close the PR
